scripts/run_perf.py

```diff
@@ -21,75 +17,268 @@
+import json
 import re
 import subprocess
+import sys
+from datetime import datetime, timezone
 from pathlib import Path
-from typing import Generator
+from typing import Any, Dict, List
+
+sys.path.append(str(Path(__file__).parent.parent))
+from savant.utils.platform import get_jetson_stats, get_platform_info
+from savant.utils.version import version
 
-# ANSI codes are used, for example, to colorize terminal output. They come from savant_rs logging
+# ANSI codes are used, for example, to colorize terminal output.
+# They come from savant_rs logging.
 # ANSI codes interfere with parsing FPS from the output.
 ANSI_ESCAPE_PATTERN = re.compile(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])')
+FPS_PATTERN = re.compile(r'^.*Processed \d+ frames, (?P<fps>\d+\.\d+) FPS\.$')
+STATS_PATTERN = re.compile(
+    r'^(?P<name>\w+): min=(?P<min>\d+), max=(?P<max>\d+), avg=(?P<avg>\d+\.\d+)$'
+)
+
+# short measurement period to drop the 1st and the last measurements (outliers)
+# TODO: Implement delayed start and early stop of fps measurements in pipeline
+FPS_PERIOD = 100
+
 
+def process_logs_file(logs_file_path: Path, stats_vars: List[str]) -> Dict[str, Any]:
+    """Get stats from log file."""
+    stats = {k: {} for k in stats_vars}
+    fps_list = []
+    with open(logs_file_path, mode='r') as log_file:
+        for line in log_file:
+            line = line.rstrip()
+            if not line:
+                continue
+
+            # checks won't work if there are ANSI codes in the line
+            line = ANSI_ESCAPE_PATTERN.sub('', line)
+
+            if ' Processed ' in line:
+                match = FPS_PATTERN.match(line)
+                if match:
+                    fps_list.append(float(match['fps']))
+                continue
+
+            for stats_var in stats_vars:
+                if stats_var in line:
+                    match = STATS_PATTERN.match(line)
+                    if match:
+                        stats[match['name']]['min'] = int(match['min'])
+                        stats[match['name']]['max'] = int(match['max'])
+                        stats[match['name']]['avg'] = float(match['avg'])
+                    break
 
-def launch_script(script: Path) -> Generator[str, None, None]:
-    """Runs the script.
-
-    :param script: The script.
-    :return: Output rows
-    """
-    with subprocess.Popen(
-        script,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT,
-        # env=env,
-        encoding='utf8',
-        text=True,
-    ) as process:
-        while True:
-            ret_code = process.poll()
-            line = process.stdout.readline().strip()
-            yield line
-            if process.stderr:
-                yield process.stderr.readline().strip()
-            if ret_code is not None:
-                break
+    # remove perhaps outliers
+    if len(fps_list) > 2:
+        fps_list = fps_list[1:-1]
+    # average
+    if len(fps_list) > 1:
+        num_frames = len(fps_list) * FPS_PERIOD
```

Comment:
Won't this produce incorrect `num_frames` in case the number of frames in the source isn't evenly divisible by the FPS_PERIOD?

Will it be better to get num frames directly from fps messages by capturing the `\d+` in `Processed \d+ frames`?

Reply:
You are right. I'll fix it.